### PR TITLE
Forbid to have getter functions that returns nothing (#1382)

### LINKED ISSUE #1382: `get_*` functions must return something.
# Rule request

## Thesis

Functions and methods with `get_` prefix must return something (have `return` or `yield` in the body).

## Reasoning

`get_` function is expected to return what it promises to return. For example, `get_user` returns a user object. If it's not so, it's an error or a bad interface.

### PR DESCRIPTION
**Pull request by hhsdev**
# I have made things!

- Add `GetterWithoutReturnViolation`, triggered when functions and
  methods whose name starts with `get_` returns nothing

- Rename `FunctionArgumentVisitor` to `FunctionSignatureVisitor`,
  in preparation to add code that checks the function name as well

- Add code checking for `GetterWithoutReturnViolation` to
  `FunctionSignatureVisitor`

- Add noqa tests

- Add unit test for `GetterWithoutReturnViolation` in
  `test_functions/test_getter_without_return.py`

- Update CHANGELOG.md appropriately

## Checklist

- [x] I have double checked that there are no unrelated changes in this pull request (old patches, accidental config files, etc)
- [x] I have created at least one test case for the changes I have made
- [x] I have updated the documentation for the changes I have made
- [x] I have added my changes to the `CHANGELOG.md`

- Closes #1382

### CODE REVIEW

File: tests/test_visitors/test_ast/test_functions/test_getter_without_return.py [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+from wemake_python_styleguide.violations.best_practices import (
+    GetterWithoutReturnViolation,
+)
+from wemake_python_styleguide.visitors.ast.functions import (
+    FunctionSignatureVisitor,
+)
+
+getter_function_with_implicit_return = """
+def get_foo():
+    ...
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_functions/test_getter_without_return.py:12]: This should be allowed, because we can use WPS with stub files.

File: tests/test_visitors/test_ast/test_functions/test_getter_without_return.py [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+    code,
+    mode,
+):
+    """Testing that getters which output values are allowed."""
+    tree = parse_ast_tree(mode(code))
+
+    visitor = FunctionSignatureVisitor(default_options, tree=tree)
+    visitor.run()
+
+    assert_errors(visitor, [])
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_functions/test_getter_without_return.py:201]: Let's also test a function / method with just a docstring.
It should be allowed: intreface definitions, etc.

We should also allow `get_` function with just `raise` (+ an optional docstring). It is used in interfaces as well.

File: wemake_python_styleguide/violations/best_practices.py [hunk 3]
```diff
@@ -2255,3 +2257,47 @@ class InfiniteWhileLoopViolation(ASTViolation):
 
     error_template = 'Found an infinite while loop'
     code = 457
+
+
+class GetterWithoutReturnViolation(ASTViolation):
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/best_practices.py:2262]: Should be `@final`

File: wemake_python_styleguide/violations/best_practices.py [hunk 4]
```diff
@@ -2255,3 +2257,47 @@ class InfiniteWhileLoopViolation(ASTViolation):
...
+        A ``get_`` function is generally expected to return a value. Otherwise,
+        it is most likely either an error or bad naming.
+
+    Solution:
+        Make sure getter functions ``return`` or ``yield`` a value on all
+        execution paths, or rename the function.
+
+    Example::
+
+        # Correct:
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/best_practices.py:2278]:
```python
# Correct
def get_random_number():
     return random.randint(1, 10)

# Wrong
def get_random_number():
     print('I do not return a value!')
```

File: wemake_python_styleguide/violations/best_practices.py [hunk 5]
```diff
@@ -2255,3 +2257,47 @@ class InfiniteWhileLoopViolation(ASTViolation):
...
+
+            def get_data_into_database(self):  # load or update is better
+                self.database.update(http.request(url))
+
+
+    .. versionadded:: 0.15.0
+
+    """
+
+    error_template = 'Found getter function without a return value'
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/best_practices.py:2302]: `Found a getter without a return value`

File: wemake_python_styleguide/visitors/ast/functions.py [hunk 6]
```diff
@@ -1,6 +1,6 @@
 import ast
+import typing
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/ast/functions.py:2]: Please, don't use `import typing`
Because typing is already verbose, extra `typing.` makes it even worse 😢

**Review by sobolevn — CHANGES_REQUESTED** [review]
Thanks a lot! Let's add some more test to be sure that we have covered every possible corner case.

File: wemake_python_styleguide/visitors/ast/functions.py (hunk 6 above)
Comment (hhsdev) [reply wemake_python_styleguide/visitors/ast/functions.py:2]: I was bugged about it as well, but it's saying too many imported members (10 > 8)

Extra types imported by me were `Type`, `Tuple` and `Iterator`, all needed for `_get_function_exit_nodes`. I thought about using `walk.get_subnodes_by_type` instead, but that function doesn't seem to take tuples as type.

Comment (sobolevn) [reply wemake_python_styleguide/visitors/ast/functions.py:2]: Maybe we can extract some parts to `logic/`?

**Comment by sobolevn** [comment]
Btw, there are some merge conflicts. Please, rebase your build 👍

File: tests/test_visitors/test_ast/test_functions/test_getter_without_return.py (hunk 2 above)
Comment (hhsdev) [reply tests/test_visitors/test_ast/test_functions/test_getter_without_return.py:201]: Do we also allow if the `get_` function has extra statements besides the raise statement?
```
def get_foo():
    print('Debug: get_foo is called')
    raise ValueError('error!')
```

Comment (sobolevn) [reply tests/test_visitors/test_ast/test_functions/test_getter_without_return.py:201]: Nope, only a single `raise`

File: tests/test_visitors/test_ast/test_functions/test_getter_without_return.py [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+"""
+
+
+@pytest.mark.parametrize('code', [
+    getter_function_with_implicit_return,
+    getter_function_with_bare_return,
+    getter_function_with_bare_yield,
+    getter_method_with_implicit_return,
+    getter_method_with_bare_return,
+    getter_method_with_bare_yield,
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_functions/test_getter_without_return.py:159]: Bare `yield` should be allowed. Because `yield` can be used to define different constructs like `contextmanager` and `fixture`.

Can we please add this code as a test:

```python
def wrapper():
    print('context manager')
    yield
```

This should be allowed.

File: wemake_python_styleguide/visitors/ast/functions.py [hunk 8]
```diff
@@ -432,21 +453,53 @@ class FunctionArgumentsVisitor(base.BaseNodeVisitor):
...
+        has_explicit_function_exit = False
+        for function_exit_node in self._get_function_exit_nodes(node):
+            has_explicit_function_exit = True
+            if function_exit_node.value is None:
+                self.add_violation(GetterWithoutReturnViolation(node))
+
+        if not has_explicit_function_exit:
+            self.add_violation(GetterWithoutReturnViolation(node))
+
+    def _get_function_exit_nodes(
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/ast/functions.py:494]: We can move this to `logic/`

**Review by sobolevn — CHANGES_REQUESTED** [review]
Thanks a lot! There are some other corner-cases that I found.

File: tests/test_visitors/test_ast/test_functions/test_getter_without_return.py (hunk 7 above)
Comment (hhsdev) [reply tests/test_visitors/test_ast/test_functions/test_getter_without_return.py:159]: that would be allowed because the function name doesn't start with `get_`. should I add the same function with name `get_wrapper`?

Comment (sobolevn) [reply tests/test_visitors/test_ast/test_functions/test_getter_without_return.py:159]: Yeap 👍

File: tests/test_visitors/test_ast/test_functions/test_getter_without_return.py [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+"""
+
+getter_function_with_yield_from = """
+def get_foo():
+    yield from [1]
+"""
+
+getter_method_with_implicit_return = """
+class Foo:
+    def get_foo():
```
Comment (skarzi) [inline tests/test_visitors/test_ast/test_functions/test_getter_without_return.py:52]: NIP: please, let's add `self` argument to all methods - it doesn't change anything, but method without `self` argument or `staticmethod` decorator is confusing

Comment (sobolevn) [reply tests/test_visitors/test_ast/test_functions/test_getter_without_return.py:52]: Good catch 👍

**Review by sobolevn — APPROVED** [review]
Thanks a lot, awesome feature!
